base/array.jl

```diff
@@ -2325,7 +2325,34 @@ julia> filter(isodd, a)
  9
 ```
 """
-filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
+function filter(f, a::Array{T, N}) where {T, N}
+    j = 1
+    b = Vector{T}(undef, length(a))
+    for ai in a
+        @inbounds b[j] = ai
+        j = ifelse(f(ai), j+1, j)
+    end
+    resize!(b, j-1)
+    sizehint!(b, length(b))
+    b
+end
+
+function filter(f, a::AbstractArray)
+    (IndexStyle(a) != IndexLinear()) && return a[map(f, a)::AbstractArray{Bool}]
+
+    j = 1
+    idxs = Vector{Int}(undef, length(a))
+    for idx in eachindex(a)
+        @inbounds idxs[j] = idx
+        ai = @inbounds a[idx]
+        j = ifelse(f(ai), j+1, j)
+    end
+    resize!(idxs, j-1)
+    res = a[idxs]
+    empty!(idxs)
+    sizehint!(idxs, 0)
```

Comment:
Does this really improve performance in the end? This breaks with `AbstractArray`s for which `getindex` returns a view which uses `idxs` to store its indices. It's not clear whether the `AbstractArray` interface allows that or not. See https://github.com/JuliaData/DataFrames.jl/pull/3192.

Reply:
@nalimilan, maybe open an issue so this doesn't get lost?

Reply:
Sure: https://github.com/JuliaLang/julia/issues/47078